internal/controller/ctrlpkg/util.go

```diff
@@ -1,7 +1,29 @@
 package ctrlpkg
 
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+)
+
 func IsSameResource(a, b Package) bool {
 	return a.GetName() == b.GetName() &&
 		a.GroupVersionKind() == b.GroupVersionKind() &&
 		a.GetNamespace() == b.GetNamespace()
 }
+
+func SpecHash(pkg Package) (string, error) {
+	if specBytes, err := json.Marshal(pkg.GetSpec()); err != nil {
+		return "", fmt.Errorf("failed to marshal package spec: %w", err)
+	} else {
+		var currentSpecHash string
+		h := sha256.New()
+		if _, err := h.Write(specBytes); err != nil {
```

Comment:
A slightly more efficient way to do this would be a streaming JSON encoder like `json.NewEncoder(h).Marshal(pkg.GetSpec())`. 

It probably wouldn't matter here performance wise, but I think it's more concise too, so I would personally prefer that way :slightly_smiling_face: